lib/semantics/mod-file.cc

```diff
@@ -293,7 +303,13 @@ std::vector<const Symbol *> CollectSymbols(const Scope &scope) {
     }
   }
   std::sort(sorted.begin(), sorted.end(), [](const Symbol *x, const Symbol *y) {
-    return x->name().begin() < y->name().begin();
+    bool xIsNml{x->has<NamelistDetails>()};
+    bool yIsNml{y->has<NamelistDetails>()};
+    if (xIsNml != yIsNml) {
+      return xIsNml < yIsNml;
```

Comment:
One doesn't often see comparisons of Booleans.

Consider `return yIsNml;`.  It'll do the same thing and might be more clear.  (Hard to tell, it's subjective.)

Reply:
What I really want is a template that combines two Compare functions into one. I.e. compare the isNml property and then the name location. But I couldn't find anything like that in the C++ standard library.

That way of thinking about it is what led me to this implementation. I see that `return yIsNml;` is equivalent, but I had to stop and think about it.

Does anyone else have an opinion on which is clearer?

Reply:
Maybe wrap the sequence in a function with naming that conveys that primary and secondary keys are being used.

```
template<typename PRIMARY, typename SECONDARY>
bool IsLess(PRIMARY x1, SECONDARY x2, PRIMARY y1, SECONDARY y2) {
  return x1 != y1 ? x1 < y1 : x2 < y2;
}
```